### PR TITLE
TINY-7399: Context menu not closing when clicking into a different editor

### PR DESCRIPTION
**Pull request by diogo405**
Related Ticket:

Description of Changes:
* Close the context menu when clicking into another editor
* Close the editor menu when clicking into another editor or opening a menu in another editor
* Add functions to `TinyContentActions` to click into the editor: `trueClick` and `trueClickOn`

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/themes/silver/main/ts/Events.ts [hunk 1]
```diff
@@ -42,10 +50,15 @@ const setup = (editor: Editor, mothership: Gui.GuiSystem, uiMothership: Gui.GuiS
   });
 
   // Editor content events
-  const onContentClick = (raw: UIEvent) => broadcastOn(Channels.dismissPopups(), { target: SugarElement.fromDom(raw.target as Node) });
+  const onContentClick = () => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/Events.ts:53]: This logic needs to be on `mousedown`, not on `mouseup`/`click`. If you recall when we were discussing this, I showed that we'd need a new `onContentMousedown` function to handle that case.

- [ ] Revert this and add a new `onContentMousedown` handle please

File: modules/tinymce/CHANGELOG.md [hunk 2]
```diff
@@ -51,6 +51,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Links in notification text did not show the correct mouse pointer #TINY-7661
 - The formatter match APIs were incorrectly returning false for formats that specified an attribute or style should be removed from an element #TINY-6149
 - The type signature on `editor.formatter.matchNode` had the wrong return type (was `boolean` but should have been `Formatter | undefined`) #TINY-6149
+- Context menu not closing when clicking into a different editor #TINY-7399
+- Menu not closing when clicking into a different editor or opening other editor menu #TINY-7399
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:55]:
```suggestion
- Menus and context menus were not closed when clicking into a different editor #TINY-7399
```
This only needs a single changelog entry.

File: modules/mcagar/src/main/ts/ephox/mcagar/api/bdd/TinyContentActions.ts [hunk 3]
```diff
@@ -19,10 +19,18 @@ const keystroke = (editor: Editor, keyvalue: number, modifiers: Keyboard.KeyModi
 const type = (editor: Editor, content: string): void =>
   TypeText.typeContentAtSelection(TinyDom.document(editor), content);
 
+const trueClick = (editor: Editor): void =>
```
Comment (lnewson) [inline modules/mcagar/src/main/ts/ephox/mcagar/api/bdd/TinyContentActions.ts:22]: These new functions need to be documented in the McAgar changelog.

- [ ] McAgar changelog entries please

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  }, [ Theme ]);
+
+  const hook2 = TinyHooks.bddSetup({
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'link',
+  }, [ Theme ]);
+
+  const waitToOpen = (selector: string) => {
+    return Waiter.pTryUntil(`Wait ${selector} to open`, () => UiFinder.exists(SugarBody.body(), selector));
+  };
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:22]:
```
  const pWaitToOpen = (selector: string) =>
    Waiter.pTryUntil(`Wait for a menu to open`, () => UiFinder.exists(SugarBody.body(), selector));
```
Shorthand and we don't need to include the selector as it'll be in the assertion error message. We also have a convention to prefix test functions that are async with `p` currently.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  const waitToClose = (selector: string) => {
+    return Waiter.pTryUntil(`Wait ${selector} to close`, () => UiFinder.notExists(SugarBody.body(), selector));
+  };
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:26]:
```
  const pWaitToClose = (selector: string) =>
    Waiter.pTryUntil(`Wait for all menus to close`, () => UiFinder.notExists(SugarBody.body(), selector));
```
Same as last time here.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  it('TINY-7399: Clicking on the editor should close the context menu', async () => {
+    const editor = hook1.editor();
+    TinyContentActions.trueClick(editor);
+    await TinyUiActions.pTriggerContextMenu(editor, 'p', '[role="menu"]');
+    TinyContentActions.trueClick(editor);
+    await waitToClose('[role="menu"]');
+  });
+
+  it('TINY-7399: Clicking on the editor should close other editor menu', async () => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:44]:
```suggestion
  it('TINY-7399: Clicking on the editor should close other editors menus', async () => {
```

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    const editor1 = hook1.editor();
+    const editor2 = hook2.editor();
+
+    TinyUiActions.clickOnMenu(editor1, '[role="menuitem"]');
+    await waitToOpen('[role="menu"]');
+    TinyContentActions.trueClick(editor2);
+    await waitToClose('[role="menu"]');
+  });
+
+  it('TINY-7399: Clicking on the editor should close other editor context menu', async () => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:54]:
```suggestion
  it('TINY-7399: Clicking on the editor should close other editors context menus', async () => {
```

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    const editor1 = hook1.editor();
+    const editor2 = hook2.editor();
+
+    TinyUiActions.clickOnMenu(editor1, '[role="menuitem"]');
+    await waitToOpen('[role="menu"]');
+    Mouse.trueClickOn(TinyDom.container(editor2), '[role="menuitem"]');
+    await waitToOpen('[role="menu"]');
+
+    const menus = UiFinder.findAllIn(SugarBody.body(), '[role="menu"]');
+    assert.equal(menus.length, 1, 'Should have one menu open');
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:73]:
```suggestion
    assert.lengthOf(menus, 1, 'Should have one menu open');
```

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+describe('browser.tinymce.themes.silver.editor.EventsTest', () => {
+  const hook1 = TinyHooks.bddSetup({
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'link',
+  }, [ Theme ]);
+
+  const hook2 = TinyHooks.bddSetup({
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'link',
+  }, [ Theme ]);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:18]:
```suggestion
  const settings = {
    base_url: '/project/tinymce/js/tinymce',
    plugins: 'link',
    menubar: 'file'
  };

  const hook1 = TinyHooks.bddSetupLight(settings, [ Theme ]);
  const hook2 = TinyHooks.bddSetupLight(settings, [ Theme ]);
```
This removes the duplication and also saves us having to render UI components we're not using in the tests.

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts (hunk 4 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:21]: Yeah, but if remove the selector from the message and pass another thing that's not a menu is not gonna sound good. Unless the method is `pWaitMenuToOpen` without params...

Comment (lnewson) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:21]: That's all you're using it for though... unless this is moved elsewhere to be generic, then that's not something we need to worry about here. You're right it should be renamed though.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  const pWaitMenuToOpen = () =>
+    Waiter.pTryUntil('Wait for a menu to open', () => UiFinder.exists(SugarBody.body(), menuSelector));
+
+  const pWaitMenuToClose = () =>
+    Waiter.pTryUntil('Wait for all menus to close', () => UiFinder.notExists(SugarBody.body(), menuSelector));
+
+  it('TINY-7399: Clicking on the editor should close the menu', async () => {
+    const editor = hook1.editor();
+    TinyUiActions.clickOnMenu(editor, '[role="menuitem"]');
```
Comment (ltrouton) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts:28]: Could be worth making `'[role="menuitem"]'` a constant like `menuSelector`.

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    await pWaitMenuToOpen();
+    TinyContentActions.trueClick(editor);
+    await pWaitMenuToClose();
+  });
+
+  it('TINY-7399: Clicking on the editor should close the context menu', async () => {
+    const editor = hook1.editor();
+    TinyContentActions.trueClick(editor);
+    await TinyUiActions.pTriggerContextMenu(editor, 'p', '[role="menu"]');
```
[the remark this answers is not in the document]
Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts]: Yes!

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts (hunk 10 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/EventsTest.ts]: Created the constant, thanks for spotting that 👍

**Review by ashfordneil — APPROVED** [review]

**Pull request merged**